src/mds/CDir.h

```diff
@@ -169,22 +169,35 @@ class CDir : public MDSCacheObject, public Counter<CDir> {
 
   CInode          *inode;  // my inode
   frag_t           frag;   // my frag
+  snapid_t	   first;
+
+  using fnode_ptr = std::shared_ptr<fnode_t>;
+  using fnode_const_ptr = std::shared_ptr<const fnode_t>;
+
+  template <typename ...Args>
+  static fnode_ptr allocate_fnode(Args && ...args) {
+    static mempool::mds_co::pool_allocator<fnode_t> allocator;
+    return std::allocate_shared<fnode_t>(allocator, std::forward<Args>(args)...);
+  }
+
+  static fnode_ptr const_fnode_cast(const fnode_const_ptr& ptr) {
```

Comment:
This method looks evil. Why do we need to cast from `fnode_const_ptr` to `fnode_ptr`?

Reply:
it's for CDir::_get_fnode() and CDir::_get_projected_fnode(). _get_fnode() is for modifying fnode of newly allocate CDir. _get_projected_fnode() is for functions that know fnode is already projected and wants to modify it. They are only used in limited place.